include/rocksdb/utilities/object_registry.h

```diff
@@ -38,6 +39,30 @@ using FactoryFunc =
 // library
 using RegistrarFunc = std::function<int(ObjectLibrary&, const std::string&)>;
 
+// The signature of the function for loading plugins
+// The Plugin struct is passed to the PluginFunc and expected to be validated
```

Comment:
Like validating the version field? This seems like an obscure way of having some input parameters and some output parameters without making it clear which are which.

Reply:
> This code still seems very haphazard and arbitrary. The tests don't give me many clues about what practical use will look like.

There are a few main use cases for plugins:
1.  Registering extensions to RocksDB at build time.  This is currently built into the Make/CMake system.  Plugins define their PluginFunc that fills in the properties about it (name, registration func/arg).  When a RocksDB process starts, the build-time plugins will be registered automatically.  This methodology is portable across Make/CMake and different operating systems (unlike the current methodology that only works on Linux).
2. Registering extensions to RocksDB at run-time.  This methodology is not likely to be that useful but still exists and is what can be easily tested.  In this mode the Plugin is registered by the executable during startup.  This mode is not that useful as it could already be accomplished by using the ObjectRegistry directly.
3. Registering extensions to RocksDB dynamically.  Facebook does not want to support this option but I think it is useful to the rest of the world and may be required for Java developers.  In this mode the user would dynamically load a shared library, lookup the PluginFunc by name in the library, and invoke it.  
4. Registering extensions to RocksDB at configure time.  This model is very similar to the dynamic model but is controlled more by applications/strings/options.  In this model, we can register the name of extension libraries and their plugin functions somewhere and load those functions and libraries at configuration time (presumably as part of something like LoadLatestOptions or something similar).

More infrastructure was added to support the dynamic/configuration use cases that were deemed "not wanted" by Facebook (though I have no idea how a standard Java release works without it).  Some pieces (like not passing classes around -- and returning Status) were left to prevent some ABI-related issues.

To work around some of the potential ABI-issues, I had three versions (the RocksDB version that the system was compiled with; the RocksDB version that the plugin was compiled with; and the version of the Plugin struct itself.  The first two were to allow the Plugin and RocksDB to check for compatibility and the latter was for ABI with the Plugin struct.  All have since been removed in the hope of making progress on some part of this PR.

Note that Facebook already has its own implementation of a the "PluginFunc" -- it is called "RegisterCustomObjects" and is used to inject FB-specific code into the unit tests.  The plugin infrastructure will allow the "RegisterCustomObjects" to be deleted and replaced by compile-time plugins.  Furthermore, the code that is used in the FB applications that does the equivalent of "RegisterCustomObjects" can also be removed.